boltdesignsystem/bolt
# Globally wire up remaining Bolt components to use new lazy queue / on-demand loading pattern

## Jira
http://vjira2:8080/browse/BDS-2105

## Dependencies**
**to avoid any rendering issues and keep the # of changes to a minimum

- ~[Migrate Accordion to LitElement](https://github.com/boltdesignsystem/bolt/pull/1858)~
- ~[Add Classes to Slotted Children](https://github.com/boltdesignsystem/bolt/pull/1859) decorator work~
- ~[Migrate components to use `<ssr-keep>`](https://github.com/boltdesignsystem/bolt/pull/1857)~
- ~Switch from _eagerly_ loading components to [lazily loading components](https://github.com/boltdesignsystem/bolt/pull/1854)~

## Summary
In a nutshell, this PR takes the new lazy queue web performance updates from https://github.com/boltdesignsystem/bolt/pull/1821 and rolls it out to our components globally -- which should result in _significantly_ improved client-side performance.

For example, with these updates (plus the two separately PR'd updates mentioned below), our main JS bundle goes from being around **`1.9MB`** in size (minified but not gzipped) down to around **`50KB`** (again, minified but not gzipped).

![image](https://user-images.githubusercontent.com/1617209/79868536-0a628b80-83ae-11ea-8733-721919597032.png)

> Even unoptimized, these updates start to show the full perf improvements seen on the full POC branch (which tests out all of these updates, combined)

<hr>

## Performance Impact
- All dynamic imports in Bolt are now bundled and loaded separately ("lazy") vs eagerly bundled together as they were previously
- All JS bundles in Bolt are now loaded with the new lazy queue package added in #1821 -- this means that only the components required on a page are loaded and components that get injected later will load on demand
- When compiled for production (without making any local code changes), the main `bolt-global.modern.js` JS file size should now be around `360KB` (minified but not gzipped)

Note: once the work from https://github.com/boltdesignsystem/bolt/pull/1855 and https://github.com/boltdesignsystem/bolt/pull/1846 get in, the fully compiled JS from Bolt (the initial JS that gets loaded) should be closer to 50KB (minified) or around 14KB (minified + gzipped)!

## How To Test
As weird as this might sound, I'd actually recommend we don't focus too much on the actual code for this (at least initially) and initially focus our efforts on spot checking the staged docs site.

Besides confirming that our tests are all passing (which is super easy to do), the main thing that we need to do here is just poke around Pattern Lab to do another across-the-board gut check (especially on heavier demos like some of the Academy page Blueprints or the Tabs / Accordion demos) and verify that things all seem to be working exactly as expected.

Components are loading fine, nothing is rendering weirdly, no flashes of unstyled UI, etc.

Some cross-browser spot checks would also be wise but overall I'm not expecting many (if any) surprises!
